// @lyrasearch/lyra/dist/cjs/src/prefix-tree/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeDocument = exports.updateParent = exports.create = void 0;
const utils_1 = require("../utils");
function create(key = "") {
    const node = {
        id: (0, utils_1.uniqueId)(),
        key,
        word: "",
        parent: null,
        children: {},
        docs: [],
        end: false,
    };
    Object.defineProperty(node, "toJSON", { value: serialize });
    return node;
}
exports.create = create;
function updateParent(node, parent) {
    node.parent = parent.id;
    node.word = parent.word + node.key;
}
exports.updateParent = updateParent;
function removeDocument(node, docID) {
    const index = node.docs.indexOf(docID);
    /* c8 ignore next 3 */
    if (index === -1) {
        return false;
    }
    node.docs.splice(index, 1);
    return true;
}
exports.removeDocument = removeDocument;
/* c8 ignore next 5 */
function serialize() {
    const { key, word, children, docs, end } = this;
    return { key, word, children, docs, end };
}
